Modeling Dynamic Coating for Opaque Surfaces

yujiex opened the pull request:
Pull request overview
---------------------
Currently, EnergyPlus assumes constant thermal and solar absorptance for exterior wall and roof materials. In order to model dynamic coatings with changing solar or thermal absorptance, users need to use the EnergyManagementSystem (EMS) feature to write code to overwrite these properties. Although this can be done by advanced users, it can be a challenge for normal EnergyPlus users as learning and debugging EMS code is not an easy task. 

The proposed new feature will enable user input of the stimulus-absorptance relationship in the form of a table or curve. The variables triggering solar or thermal absorptance change can be surface temperature, solar radiation, space heating/cooling status, a schedule, etc. An idd object will be added to capture detailed input of solar and thermal absorptance as a function of different stimuli.

The PR also includes a refactoring of the `MaterialProperties` class into a parent `Material` class and then child classes `MaterialNoMass`, `MaterialGas`, etc. that have only the fields required for the different types of materials. This avoids the allocation of big chunks of memory while only a handful of fields are in use.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex committed:
initial commit

yujiex committed:
Add newline to citation items

yujiex committed:
Add physical newline to citations

yujiex committed:
revert to add br

amirroth commented:
I'm going to attach a refactor of the `MaterialProperties` class to this feature.

yujiex commented:
> I'm going to attach a refactor of the `MaterialProperties` class to this feature.

Sure. I will also work on the refactoring of the `MaterialProperties` class in this PR.

amirroth commented:
It may be better to do this cleanup before you get started on implementing the new feature, so that the new feature can be implemented in the clean environment. 

The basic idea is to break the `MaterialProperties` class into a parent `Material` class and then child classes `MaterialNoMass`, `MaterialGas`, etc. that have only the fields required for the different types of materials, so that you are not allocating these 1KB structures (not even counting embedded arrays) and then leaving most of them unused, which destroys memory locality. The `Material` array would become an array of pointers to `Material` class parent objects. You would then implement the new feature as a new subclass `class MaterialDynamicCoating : public Material`. A further potential cleanup could be to update the getInput routine to use epJSON input.

yujiex commented:
> It may be better to do this cleanup before you get started on implementing the new feature, so that the new feature can be implemented in the clean environment.
> 
> The basic idea is to break the `MaterialProperties` class into a parent `Material` class and then child classes `MaterialNoMass`, `MaterialGas`, etc. that have only the fields required for the different types of materials, so that you are not allocating these 1KB structures (not even counting embedded arrays) and then leaving most of them unused, which destroys memory locality. The `Material` array would become an array of pointers to `Material` class parent objects. You would then implement the new feature as a new subclass `class MaterialDynamicCoating : public Material`. A further potential cleanup could be to update the getInput routine to use epJSON input.

Got you. I will work on breaking the `MaterialProperties` class into subclasses first.

amirroth commented:
The way I would do this is:
1.  Change the `Materials` array to an `EPVector<MaterialProperties *>`. This will require a bunch of syntactic changes from `.` to `->`, but that's it. 
2. Split the `MaterialProperties` class into a `MaterialBase` class which contains `Name`, `Group`, and maybe a few other things, I'm not sure and a `MaterialChild` class which contains everything else. Change `Materials` array to an `EPVector<MaterialBase *>`. This will require you to add `MaterialChild *thisMaterialChild = dynamic_cast<MaterialChild *>state.dataMaterial->Materials(MaterialNum);` local variables in a bunch of places, but again nothing else.
3. Now you can start splitting `MaterialChild` into smaller classes, one class at a time, until `MaterialChild` is the final remaining class, whatever it is.  

This is an incremental, testable approach to making this kind of transition.

yujiex commented:
> The way I would do this is:
> 
> 1. Change the `Materials` array to an `EPVector<MaterialProperties *>`. This will require a bunch of syntactic changes from `.` to `->`, but that's it.
> 2. Split the `MaterialProperties` class into a `MaterialBase` class which contains `Name`, `Group`, and maybe a few other things, I'm not sure and a `MaterialChild` class which contains everything else. Change `Materials` array to an `EPVector<MaterialBase *>`. This will require you to add `MaterialChild *thisMaterialChild = dynamic_cast<MaterialChild *>state.dataMaterial->Materials(MaterialNum);` local variables in a bunch of places, but again nothing else.
> 3. Now you can start splitting `MaterialChild` into smaller classes, one class at a time, until `MaterialChild` is the final remaining class, whatever it is.
> 
> This is an incremental, testable approach to making this kind of transition.



> 1. EPVector<MaterialProperties *>

Thanks for the tips. This is a nicer way to go. I'll follow this method to make the changes.

JasonGlazer commented:
@yujiex a few comments: 
- Are you planning on including Roughness, Thickness, Conductivity, Density, and Specific heat in your input object? Or because they are "coatings" are you assuming that they are so thin that these properties are the same as the material beneath them?
- "stimuli" is a word that I don't think we are using anywhere else and maybe we should think of a better term

amirroth commented:
I should probably add that I don't have any comments about how the feature is intended to work. I guess one question is whether the EMS functionality is being retained or deprecated in favor of this feature?

mjwitte commented:
@yujiex On a prior call, we discussed the option of making this `MaterialProperty:DynamicCoating` which would modify a base material object, similar to `MaterialProperty:PhaseChange` and `MaterialProperty:VariableThermalConductivity` Did you decide against that approach? If so, a few comments in the NFP about that would be welcome.

yujiex commented:
> * Are you planning on including Roughness, Thickness, Conductivity, Density, and Specific heat in your input object? Or because they are "coatings" are you assuming that they are so thin that these properties are the same as the material beneath them?




> @yujiex a few comments:
> 
> * Are you planning on including Roughness, Thickness, Conductivity, Density, and Specific heat in your input object? Or because they are "coatings" are you assuming that they are so thin that these properties are the same as the material beneath them?
> * "stimuli" is a word that I don't think we are using anywhere else and maybe we should think of a better term

@JasonGlazer 
These materials could be coatings or some films. The Roughness could be different from the layer below it, as a lot of these materials are highly reflective, and probably will be smoother. The thickness is probably thin for a lot of them, although some are like a compound mechanical structure, which might not be as thin as a film. Should change the title of "dynamic coating" to "exterior dynamic layer" or something else to mean it could be a thicker layer?

For the "stimuli" wording, do you think "trigger" is better? I mainly want to express that these materials respond to the change of such variables.

yujiex commented:
> @yujiex On a prior call, we discussed the option of making this `MaterialProperty:DynamicCoating` which would modify a base material object, similar to `MaterialProperty:PhaseChange` and `MaterialProperty:VariableThermalConductivity` Did you decide against that approach? If so, a few comments in the NFP about that would be welcome.

Yes, as if we use `MaterialProperty:*`, it's going to look like this, where each trigger/stimuli variable will need an object like the following. Would that be a bit complicated for users? 

<img width="1174" alt="image" src="https://user-images.githubusercontent.com/13421161/204880958-5c08705e-bf16-4fb2-b3da-ceb481325de8.png">

yujiex commented:
> I should probably add that I don't have any comments about how the feature is intended to work. I guess one question is whether the EMS functionality is being retained or deprecated in favor of this feature?

The EMS functionality should probably still be retained and users can choose which way they prefer maybe? Although it might be a bit difficult for EMS when the temperature-absorptance relationship is a curve

mjwitte commented on design/FY2023/NFP-DynamicCoatingWallRoof.md:
```diff
@@ -0,0 +1,99 @@
+    Material:DynamicCoating,
+        \memo Materials with variable properties of thermal emissivity and/or solar reflectance
+        \min-fields 8
+      A1 , \field Name
+          \required-field
+          \type alpha
+          \reference MaterialName
+      …
+      N5 , \field Visible Absorptance
```
Referring to @JasonGlazer 's comment. Does the `...` imply additional fields here? If so, please include all proposed fields. If not, then this can't be a complete material layer, because it won't have any thermal properties for the CTF or FD calculations. If not, then delete the `...`.

yujiex replied:
Yes. I will add those fields in.

yujiex replied:
Just added the in.

mjwitte commented:
@yujiex I don't see why three separate property objects would be necessary as long as the name is general enough, one object should suffice.

yujiex commented:
> @yujiex I don't see why three separate property objects would be necessary as long as the name is general enough, one object should suffice.

Oh, right. Something like this? Earlier the name was Materal: instead of MaterialProperty:, I corrected it. @mjwitte 

```
    MaterialProperty:VariableThermalAbsorptance,
      A1 , \field Name
           \required-field
           \type object-list
           \object-list MaterialName
           \note Regular Material Name to which the additional properties will be added.
           \note this the material name for the basic material properties.
      A2 , \field Stimuli Indicator
           \required-field
           \type choice
           \key SurfaceTemperature
           \key SurfaceReceivedSolarRadiation
           \key SpaceHeatingCoolingMode
           \key Scheduled
           \default SurfaceTemperature
      A3 , \field Stimuli Thermal Absorptance Relationship Data Name
          \note A Curve:* or Table:Lookup object encoding the relationship between the stimuli and surface thermal absorptance. For longwave radiation.
      A4 ; \field Thermal Absorptance Schedule Name
           \note only used when Stimuli variable = “scheduled”

    MaterialProperty:VariableSolarAbsorptance,
      A1 , \field Name
           \required-field
           \type object-list
           \object-list MaterialName
           \note Regular Material Name to which the additional properties will be added.
           \note this the material name for the basic material properties.
      A2 , \field Stimuli Indicator
           \required-field
           \type choice
           \key SurfaceTemperature
           \key SurfaceReceivedSolarRadiation
           \key SpaceHeatingCoolingMode
           \key Scheduled
           \default SurfaceTemperature
      A3 , \field Stimuli Solar Absorptance Relationship Data Name
          \note A Curve:* or Table:Lookup object encoding the relationship between the stimuli and surface solar absorptance. For longwave radiation.
      A4 ; \field Solar Absorptance Schedule Name
           \note only used when Stimuli variable = “scheduled”
```

yujiex committed:
Add in the "..." fields in the idd change section of the md

yujiex committed:
Changed stimuli to trigger, added material add-on idd design

yujiex committed:
Change field name: Relationship Data Name -> Function Name

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Change to MaterialProperty:* approach, change wording of field

- chose the material add-on approach in the idd design
- field name change: trigger indicator->control signal

yujiex committed:
Add a name field for the material itself

so that the MaterialProperty:* add-on has a name for itself and a name the
material that references it.

yujiex committed:
Changed the reference material to a different name

yujiex committed:
Change Material to EPVector<MaterialProperty ptrs> for non-window

only adjusted the . to -> for non-window material in this commit

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Partially used shortcuts for material and construction

- state.dataConstruction->Construct()->Name etc. changed to thisConstruct
after defining auto &thisConstruct = state.dataConstruction->Construct()
- state.dataMaterial->Material()->Name etc. changed to thisMaterial
after defining auto *thisMaterial = state.dataMaterial->Material()

mostly worked on construction

yujiex committed:
auto scaler type to explicitly spell out the type

yujiex committed:
Rename FindPtrItemInList to FindItemInPtrList

yujiex committed:
switch case assigning bool changed to directly compute bool

yujiex committed:
replace many ! with one, add parenthesis

yujiex committed:
BITF_TEST_ANY to regular bool comparison, material related

yujiex committed:
More change to thisMaterial shortcut pattern

yujiex committed:
Move MaterialGroup from DataHeatBalance to Material

yujiex committed:
More move MaterialGroup from DataHeatBalance to Material

yujiex committed:
use getEnumerationValue for material roughness

yujiex committed:
Remove ValidateMaterialRoughness, use getEnumerationValue cont

yujiex committed:
Remove auto for scalers in src/EnergyPlus

yujiex committed:
Remove auto for scalers in third_party/Windows-CalcEngine

yujiex committed:
Remove auto for scalers in unit tests

yujiex committed:
Remove auto for scalers in other places

yujiex committed:
Use Booleanswitch with checks on empty str

yujiex committed:
Fix creation of thisConstructSh when ConstrNumSh = 0

only create thisConstructSh when it is used instead of putting it at the
beginning, which could have array1d out of range access

yujiex committed:
fix thisMaterial shortcut: recreate after idx recomputed

yujiex committed:
More using material shortcut

yujiex committed:
Revert BITF_TEST_ANY

yujiex committed:
use getEnumerationValue for GasType

yujiex committed:
Use getEnumerationValue on GapVentType

yujiex committed:
use getEnumerationValue on slatAngleType

yujiex committed:
clang-format

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
more clang-format

yujiex committed:
Move material shortcut to after the check prevent ill access

yujiex committed:
Revert back from shortcut after redimension

the data is cleared with redimension, will err

yujiex committed:
Update thisMaterial shortcut when LayPtr is changed

yujiex committed:
clang-format

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
change classname GasTypeEnum to GasType, GasType arr to gasTypes

yujiex committed:
change GasVentTypeEnum to GasVentType, variable to lower case

yujiex committed:
change SlatAngleTypeEnum to SlatAngleType, variable to lower case

yujiex committed:
Remove GasCoeffs and use GasType from Material

yujiex committed:
clang-format

yujiex committed:
const to constexpr

yujiex committed:
clang-format of the #if comments

yujiex committed:
std::string to constexpr string_view or string &

- when the downstream function needs a string &, then it's changed to str ref,
  otherwise changed to string_view
- some error message changed to use format() string rather than "+"

yujiex committed:
Changed auto &gasMaterial(*xx) to auto *gasMaterial(xx)

- changed getGas to accept pointer
- The function getLayerMaterial returns a pointer, so also changed
  getSolidLayer, getGapLayer, getComplexGapLayer to accept pointers

yujiex committed:
change auto of string to explicit string

yujiex committed:
change auto in for loop and in some string to explicit type spec

yujiex committed:
More material shortcut MatGapL, MatGapR, MatSh

yujiex committed:
Remove checker for output from getEnumerationValue

yujiex committed:
Align CasCoeffsCon etc with gasType enum, remove temp var gasTypeInt

padded the beginning of GasCoeffsCon, GasCoeffsVis, GasCoeffsCp, GasWght,
GasSpecificHeatRatio, so that using gasType to index into these arrays does not
require the "-1" shifting

yujiex committed:
change hardcoding of air idx to enum casting to int

yujiex committed:
Move blinds related function from General to WindowManager

yujiex committed:
More material shortcut

yujiex committed:
Move TotMaterials from dataHeatBal to dataMaterial

yujiex committed:
Add gapVentName field in material rather than create a temp var

yujiex committed:
clang-format

yujiex committed:
resolve upper case to title case in gapVentName

yujiex committed:
fix format string arg count issue

yujiex committed:
Remove creation of std::string when used in format string

yujiex committed:
change calculationMethod to enum class and use getEnumerationValue

yujiex committed:
add * to the output of getLayerMaterial for clarity

yujiex committed:
another place to replace string ref with string_view

yujiex committed:
change gasName to use string_view

yujiex committed:
Remove TypeOfGas temp var

yujiex committed:
use gasType shortcut

yujiex committed:
use thisBlind shortcut

yujiex committed:
Use thisBlindBack shortcut

yujiex committed:
Remove GasName field, use gasTypeNames to look up

yujiex committed:
remove temp var InterpProfAng

yujiex committed:
const to constexpr

yujiex committed:
change front declaration of local vars to when they are used

yujiex committed:
more constexpr

yujiex committed:
remove var front declaration, use std::clamp, remove redundant ifs

yujiex committed:
adding one new material, not creating all material in the vector

yujiex committed:
remove gapVentName field

yujiex committed:
Move GetMaterialData to Material.cc

yujiex committed:
Add deletion of individual elements in Material before dealloc

yujiex committed:
Move GapVentTypeNames to Material.cc

yujiex committed:
clang-format

yujiex committed:
format print fix missing arg

yujiex committed:
Change input arg to string_view in createAirMaterialFromDistance

yujiex committed:
move thisBlind init to inside so BlNum guaranteed non-zero

yujiex committed:
include EnergyPlusData.hh, remove using namespace, clang-format

yujiex committed:
Move GasCoeffsCon etc, remove Material::, change gasType ifs

yujiex committed:
Move constexpr arrays to Material.hh

yujiex committed:
clang-format

yujiex committed:
clang-format

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
change Material.cc license year

yujiex committed:
Merge 'origin/develop' and resolve conflicts

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

yujiex committed:
Revert "Remove auto for scalers in third_party/Windows-CalcEngine"

This reverts commit 4178a8a231e9c64e248da8e41cce2b62541a7231.

yujiex committed:
Revert "Remove auto for scalers in other places"

This reverts commit 494aecfa6b5dcad1442b96202dcf913b5629f437.

Myoldmopar commented:
@yujiex that's waaay better!  There are still 2 third party files that show changes, but nice work cleaning that up quickly.

yujiex committed:
more revert Window-CalcEngine changes

yujiex commented:
> @yujiex that's waaay better! There are still 2 third party files that show changes, but nice work cleaning that up quickly.

Hi @Myoldmopar, I just reverted those two places. Now it seems no third party files are changed

yujiex committed:
Fix gasName type after reverting changes in Gases::CGas

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoof

Myoldmopar commented:
@yujiex the documentation build failed for some reason.  I don't believe it's anything to do with this branch, but I went ahead and restarted it to see if it will work before merging.  The previous CI results showed a regression diff in one file, but it appears to be just due to the CI running after a different PR merged.  I'll let Mac and Linux finish up here and if it is clean I'll merge this in.

Myoldmopar commented:
Excellent, documentation is happy and CI regression returned all clean.  This is merging now, thanks @yujiex

yujiex commented:
> Excellent, documentation is happy and CI regression returned all clean. This is merging now, thanks @yujiex

Thanks, Edwin :)
